p2p/src/transport.rs

```diff
@@ -0,0 +1,129 @@
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+
+use crate::peer::{self, Peer};
+
+// TODO(xla): Use actual PeerId type.
+type PeerId = String;
+
+pub struct BindInfo {
+    pub addr: SocketAddr,
+    pub advertise_addrs: Vec<SocketAddr>,
+    pub peer_id: PeerId,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+pub trait Connection: Clone {
+    type Stream;
+
+    fn advertised_addrs(&self) -> Vec<SocketAddr>;
+    fn close(&self) -> Result<(), Error>;
+    fn open(&self) -> Result<Self::Stream, Error>;
+    fn peer_id(&self) -> PeerId;
```

Comment:
don't think underlying transport should be aware of the peer abstraction. what necessitated this?

Reply:
The `PeerId` is not part of the peer abstraction, it's the essential identity of a node. My standing assumption is that we will not be able to transform the handshake into a protocol and have to leave that up-to the transport implementations. So they will need to have access to it to exchange it. Furthermore we likely want to enable early termination based on duplicates in connected IPs/PeerIds and down the line based on deny lists in address book implementations.

Reply:
> So they will need to have access to it to exchange it.

secret connection operates on public keys, not peer ids. or peer ids == public keys? if so, that's fine

Reply:
As we established given a `PublicKey` a `PeerId` can be derived. Should there be a need to get hold of that it can be computed ont he spot. So this should indeed be `public_key`. Thanks @melekes for the clarification.